File: src/data/response.kt
/**
 * Application API responses.
 *
 * These have JS counterparts located in frontend/src/data.js
 */

package net.tkhamez.everoute.data

import net.tkhamez.everoute.Route
import java.util.Date

enum class ResponseCodes {
    AlreadyUpdated,
    AuthError,
    ConnectionStored,
    FailedToStoreData,
    LoginEsiErrorAlliance,
    LoginEsiErrorVerify,
    LoginSsoFailed,
    LoginWrongAlliance,
    MissingInput,
    SearchError,
    SearchSuccess,
    SetWaypointsSuccess,
    SetWaypointsFailure,
    SystemNotFound,
    WrongSearchTerm,
    ImportedGates,
}

data class ResponseMessage(
    var code: ResponseCodes? = null,
    var success: Boolean? = null,
    var param: String? = null
)

data class ResponseAuthUser(
    val name: String,
    val allianceName: String,
    val allianceTicker: String,
    val roles: List<String>,
    val csrfHeaderKey: String,
    val csrfToken: String,
)

data class ResponseGates(
    var code: ResponseCodes? = null,
    var ansiblexes: MutableList<MongoAnsiblex> = mutableListOf()
)

data class ResponseSystems(
    var systems: MutableList<System> = mutableListOf()
)

data class ResponseConnectedSystems(
    var connections: MutableList<ConnectedSystems> = mutableListOf()
)

data class ResponseMapConnections(
    var code: ResponseCodes? = null,
    val ansiblexes: MutableList<ConnectedSystems> = mutableListOf(),
    val temporary: MutableList<ConnectedSystems> = mutableListOf(),
)

data class ResponseGatesUpdated(
    var code: ResponseCodes? = null,
    var allianceId: Int? = null,
    var updated: Date? = null
)

data class ResponseTemporaryConnections(
    var code: ResponseCodes? = null,
    var temporaryConnections: MutableList<MongoTemporaryConnection> = mutableListOf()
)

data class ResponseSystemNames(
    var systems: MutableList<String> = mutableListOf()
)

data class ResponseRouteLocation(
    var code: ResponseCodes? = null,
    var solarSystemId: Int? = null,
    var solarSystemName: String? = null
)

data class ResponseRouteFind(
    var code: ResponseCodes? = null,
    var routes: List<List<Route.Waypoint>> = listOf()
)

File: frontend/src/response.ts
/**
 * Counterparts of backend data classes located in src/data/response.kt.
 */

/**
 * Contains a translation key from the "responseCode" group.
 */
export type ResponseMessage = {
  code: string|null,
  success: boolean|null,
  param: string|null,
}

export type ResponseAuthUser = {
  name: string,
  allianceName: string,
  allianceTicker: string,
}

export type ResponseGates = {
  code: string|null,
  ansiblexes: Array<Ansiblex>,
}

export type ResponseSystems = {
  systems: Array<System>,
}

export type ResponseConnectedSystems = {
  connections: Array<ConnectedSystems>,
}

export type ResponseMapConnections = {
  code: string|null,
  ansiblexes: Array<ConnectedSystems>,
  temporary: Array<ConnectedSystems>,
}

export type ResponseGatesUpdated = {
  code: string|null,
  allianceId: bigint|null,
  updated: string|null, // yyyy-MM-dd'T'HH:mm:ss'Z'
}

export type ResponseTemporaryConnections = {
  code: string|null,
  temporaryConnections: Array<TemporaryConnection>,
}

export type ResponseSystemNames = {
  systems: Array<string>,
}

export type ResponseRouteFind = {
  code: string|null,
  route: Array<Waypoint>,
}

export type ResponseRouteLocation = {
  code: string|null,
  solarSystemId: bigint|null,
  solarSystemName: string|null,
}

export type Ansiblex = {
  //_id: string,
  id: number,
  name: string,
  solarSystemId: bigint,
  //allianceId: bigint,
}

export type Waypoint = {
  systemId: bigint,
  systemName: string,
  targetSystem: string,
  wormhole: boolean,
  systemSecurity: number,
  connectionType: RouteType|null,
  ansiblexId: number|null,
  ansiblexName: string|null,
}

export enum RouteType {
  Stargate = 'Stargate',
  Ansiblex = 'Ansiblex',
  Temporary = 'Temporary',
}

export type TemporaryConnection = {
  system1Id: bigint,
  system1Name: string,
  system2Id: bigint,
  system2Name: string,
  characterId: bigint,
  created: string // yyyy-MM-dd'T'HH:mm:ss'Z'
}

export type System = {
  id: bigint,
  name: string,
}

export type ConnectedSystems = {
  system1: string,
  system2: string,
}
